src/mgos_imu_icm20948.c

```diff
@@ -0,0 +1,482 @@
+
+  return true;
+}
+
+bool mgos_imu_icm20948_acc_get_odr(struct mgos_imu_acc *dev, void *imu_user_data, float *odr) {
+  uint16_t div  = 0;
+
+  if(!mgos_imu_icm20948_change_bank(dev->i2c, dev->i2caddr, imu_user_data, 2)) {
+    return false;
+  }
+
+  if(!mgos_i2c_read_reg_n(dev->i2c, dev->i2caddr, MGOS_ICM20948_REG2_ACCEL_SMPLRT_DIV_1, 2, (uint8_t *)&div)) {
+    return false;
+  }
+  div = (div & 0x00ff) << 8 | (div & 0xff00) >> 8;
+
+  // ODR is computed as follows:
+  // 1.125 kHz/(1+ACCEL_SMPLRT_DIV[11:0])
+  *odr = ACC_BASE_ODR / (div + 1);
+  return *odr >= 0;
+}
+
+bool mgos_imu_icm20948_acc_set_odr(struct mgos_imu_acc *dev, void *imu_user_data, float odr) {
+  uint16_t div = 0;
+
+  if(odr <= 0) {
+    return false;
+  }
+
+  div = (ACC_BASE_ODR / odr) - 1;
+
+  if(div > 0xfff) {
+    return false;
+  }
+
+  if(!mgos_imu_icm20948_change_bank(dev->i2c, dev->i2caddr, imu_user_data, 2)) {
+    return false;
+  }
+
+  div = (div & 0x00ff) << 8 | (div & 0xff00) >> 8;
+  if(!mgos_i2c_write_reg_n(dev->i2c, dev->i2caddr, MGOS_ICM20948_REG2_ACCEL_SMPLRT_DIV_1, 2, (uint8_t *)&div)) {
+    return false;
+  }
+
+  dev->opts.odr = odr;
+  return true;
+}
+
+bool mgos_imu_icm20948_gyro_detect(struct mgos_imu_gyro *dev, void *imu_user_data) {
+  return mgos_imu_icm20948_detect(dev->i2c, dev->i2caddr, imu_user_data);
+}
+
+bool mgos_imu_icm20948_gyro_create(struct mgos_imu_gyro *dev, void *imu_user_data) {
+  struct mgos_imu_icm20948_userdata *iud = (struct mgos_imu_icm20948_userdata *)imu_user_data;
+
+  if (!dev) {
+    return false;
+  }
+
+  // Only initialize the ICM20948 if acc hasn't done so yet
+  if (!iud->accgyro_initialized) {
+    if (!mgos_imu_icm20948_accgyro_create(dev->i2c, dev->i2caddr, imu_user_data)) {
+      return false;
+    }
+    iud->accgyro_initialized = true;
+  }
+
+  if(!mgos_imu_icm20948_change_bank(dev->i2c, dev->i2caddr, imu_user_data, 2)) {
+    return false;
+  }
+
+  // GYRO_CONFIG_1: GYRO_DLPFCFG=101(12Hz); GYRO_FS_SEL=11(2000dps); GYRO_FCHOICE=1(Enable gyro DLPF);
+  // GYRO_SMPLRT_DIV: GYRO_SMPLRT_DIV=00000000;
+  mgos_i2c_write_reg_b(dev->i2c, dev->i2caddr, MGOS_ICM20948_REG2_GYRO_CONFIG_1, 0x2F);
+  mgos_i2c_write_reg_b(dev->i2c, dev->i2caddr, MGOS_ICM20948_REG2_GYRO_SMPLRT_DIV, 0x0a);
+
+  dev->scale = 2000 / 32767.0f;
+  return true;
+}
+
+bool mgos_imu_icm20948_gyro_read(struct mgos_imu_gyro *dev, void *imu_user_data) {
+  uint8_t data[6];
+
+  if (!dev) {
+    return false;
+  }
+
+  if(!mgos_imu_icm20948_change_bank(dev->i2c, dev->i2caddr, imu_user_data, 0)) {
+    return false;
+  }
+  if (!mgos_i2c_read_reg_n(dev->i2c, dev->i2caddr, MGOS_ICM20948_REG0_GYRO_XOUT_H, 6, data)) {
+    return false;
+  }
+  dev->gx = (data[0] << 8) | (data[1]);
+  dev->gy = (data[2] << 8) | (data[3]);
+  dev->gz = (data[4] << 8) | (data[5]);
+
+  return true;
+}
+
+bool mgos_imu_icm20948_gyro_get_scale(struct mgos_imu_gyro *dev, void *imu_user_data, float *scale) {
+  uint8_t fs = 0;
+
```

Comment:
Please ensure that *scale is a valid pointer before trying to write to it later (causing a crash).
if (!scale) return false;